File: backend/src/Services/MeetingService.cs
using API.Entities;
using API.Interfaces.Repositories;
using API.Types;

namespace API.Services;

public class MeetingService(IMeetingRepository meetingRepository) {

    public async Task<Page<Meeting>> GetAll(string filter, int page, int limit) {

        Page<Meeting> meetings = await meetingRepository.GetAll(filter, page, limit);

        return meetings;

    }

    public async Task<Page<Meeting>> GetMeetingsByBuilding(int buildingId, string filter, int page, int limit) {

        Page<Meeting> meetings = await meetingRepository.GetByBuildingId(buildingId, filter, page, limit);

        return meetings;

    }

    public async Task<Meeting?> GetMeetingById(int id) {

        Meeting? meeting = await meetingRepository.GetById(id);

        return meeting;

    }

    public async Task<Meeting?> CreateMeeting(int buildingId, DateTimeOffset dateTime, int length, string description) {

        Meeting meeting = new() {
            BuildingId = buildingId,
            DateTime = dateTime,
            Length = length,
            Description = description
        };

        try {
            return await meetingRepository.Create(meeting);
        } catch {
            return null;
        }

    }

    public async Task<Meeting?> UpdateMeeting(int id, int? buildingId, DateTimeOffset? dateTime, int? length, string? description) {

        Meeting? meeting = await meetingRepository.GetById(id);

        if(meeting == null) {
            return null;
        }

        meeting.BuildingId = buildingId ?? meeting.BuildingId;
        meeting.DateTime = dateTime ?? meeting.DateTime;
        meeting.Length = length ?? meeting.Length;
        meeting.Description = description ?? meeting.Description;

        try {
            return await meetingRepository.Update(meeting);
        } catch {
            return null;
        }

    }

    public async Task<Meeting?> DeleteMeeting(int id) {

        Meeting? meeting = await meetingRepository.GetById(id);

        if(meeting == null) {
            return null;
        }

        try {
            await meetingRepository.Delete(meeting);
            return meeting;
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/meeting.service.ts
import { APIError } from "@/models/api-error.models";
import { Meeting } from "@/models/meeting.model";
import { Page } from "@/models/page";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class MeetingService extends Service {

    static async getMeetingsByBuilding(buildingId : number, filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Meeting> | APIError>(`/api/meeting/building/${buildingId}`, {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createMeeting(buildingId : number, date : Date, length : number, description : string) {

        const response = await this.axios.post<Meeting | APIError>("/api/meeting", {
            buildingId,
            date,
            length,
            description
        });

        return response.data;

    }

    static async updateMeeting(id : number, buildingId : optional<number>, dateTime : optional<Date>, length : optional<number>, description : optional<string>) {

        const response = await this.axios.put<Meeting | APIError>(`/api/meeting/${id}`, {
            buildingId,
            dateTime,
            length,
            description
        });

        return response.data;

    }

    static async deleteMeeting(id : number) {

        const response = await this.axios.delete<Meeting | APIError>(`/api/meeting/${id}`);

        return response.data;

    }

}